### ISSUE TITLE
Result variation from run to run 

### ISSUE THREAD

**Issue by lioujheyu** [mgawan/GPU-BSW#13]
Hi there,

I am conducting a research project that uses some kind of automatic tool to dynamically modify the GPU kernel for better performance. This requires the static result to be generated from the program. Initially, I was using `top_scores` as the sole metric to judge whether the newly generated GPU kernel can produce exactly the same result from the unmodified one. Recently I switch to use all output from the kernel, namely `query_begin`, `query_end`, `ref_begin`, `ref_end`, alongside with top_scores. However, I discovered that the current kernel will generate a slightly different output in `query_begin` and `ref_begin`. I was wondering if this is intended by design? Is #10 a related issue in here?

Sorry to ask a dumb question in advance as I am no expert in DNA sequence alignment

**Comment by mgawan** [comment]
Hi, thanks for your interest in our work and it looks like a very interesting project. I don't think issue 10 is related because those synchs are right at the end of kernel and should not alter results. Typically in alignment there can be multiple correct answers but I would expect this kernel to choose the same answer provided that everything else is unchanged (input scores, sequences etc). To debug/better understand this, can you please provide an example of input which gives different results across runs? are you using one of the example datasets?

Thanks

**Comment by lioujheyu** [comment]
I was using the built-in testing dataset (dna-query.fasta, dna-reference.fasta) for DNA alignment, with modification made to the program only to dump sequences (`query_begin` and so on I think) to the file. I notice that only `query_begin` and `ref_begin` differ across run, where these 2 are modified from the reverse kernel. Just FYI.

**Comment by mgawan** [comment]
so I just did multiple runs (4) and compared (diff-ed) the outputs (containing scores, start and end positions), I did not notice any discrepancies. However, there are some instructions in kernels which might give non-deterministic output on some older NVIDIA devices (I haven't confirmed this with experiments), If this turns out to be the case, I can make changes in code.

Could you please share what GPU you are using, how many GPUs are present in total on your system and what CUDA version you are building this with? Thanks.

**Comment by lioujheyu** [comment]
I have 1 Nvidia Tesla P100 in the system, with CUDA 11.1 installed.
```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 455.45.01    Driver Version: 455.45.01    CUDA Version: 11.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla P100-PCIE...  On   | 00000000:02:00.0 Off |                    0 |
| N/A   34C    P0    25W / 250W |      4MiB / 16280MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
```

**Comment by mgawan** [comment]
I just tested on a P100 device and I can now see the variation across runs, this is as suspected because of that piece of code in kernel that was written keeping in mind only the volta and post volta GPUs. I will push a fix possibly by tonight and let you know.

**Comment by mgawan** [comment]
Hi, I have been trying to debug this issue, even after removing volta specific code the results on that particular dataset (they seem to vary for a handful of sequences) do not seem to be consistent. If I remove those handful of sequences the results are consistent on P100. Which is interesting because the same code when run on AMD GPUs (with HIP) gives consistent results (including the problematic 200 sequences) and those GPUs also do not have volta specific features. I have uploaded a new dna data set in `additional_data` folder, the kernel gives persistent results if you use that dataset. I have tried this on multiple other datasets and the results seem to be persistent, I will further investigate this and update here. Meanwhile you can try out the new dataset.

**Comment by lioujheyu** [comment]
Interesting indeed. Thanks for the new dataset. I will try it out.

**Comment by mgawan** [comment]
Hi, I think I might have found the problem, the fix (b1f5905ecc8658f78cf5aed0277d7d8fc34de11a) seems to be working on our local P100 machine, the results are consistent across multiple runs on the older dataset as well. Can you please do a pull on master branch and test out the new changes on your side? Thanks.
